subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c

```diff
@@ -405,9 +735,250 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+			*id = ACCOUNT_KEY_METADATA_FIELD_GET(account_key_metadata[i], ID);
+			return 0;
+		}
+	}
+
+	return -ESRCH;
+}
+
+static struct fp_bond *fp_bond_free_get(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		if (bt_addr_le_eq(&fp_bonds[i].nv.addr, BT_ADDR_LE_ANY)) {
+			return &fp_bonds[i];
+		}
+	}
+
+	return NULL;
+}
+
+static struct fp_bond *bonded_duplicate_get(const void *conn_ctx, const bt_addr_le_t *addr)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		if ((fp_bonds[i].conn_ctx != conn_ctx) &&
+		    fp_bonds[i].bonded &&
+		    bt_addr_le_eq(&fp_bonds[i].nv.addr, addr)) {
+			return &fp_bonds[i];
+		}
+	}
+
+	return NULL;
+}
+
+static void bond_update_addr(const void *conn_ctx, const bt_addr_le_t *new_addr)
+{
+	int err;
+	struct fp_bond *bond;
+	struct fp_bond bond_rollback_copy;
+
+	bond = fp_bond_get(conn_ctx);
+	if (!bond) {
+		return;
+	}
+	bond_rollback_copy = *bond;
+
+	bond->nv.addr = *new_addr;
+
+	err = fp_bond_settings_save(bond, &bond_rollback_copy);
+	if (err) {
+		return;
+	}
+}
+
+static struct fp_bond *bonded_by_addr_get(const bt_addr_le_t *addr)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		if (fp_bonds[i].bonded &&
+		    bt_addr_le_eq(&fp_bonds[i].nv.addr, addr)) {
+			return &fp_bonds[i];
+		}
+	}
+
+	return NULL;
+}
+
+void fp_storage_ak_bond_bt_request_cb_register(const struct fp_storage_ak_bond_bt_request_cb *cb)
+{
+	if (!IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)) {
+		__ASSERT_NO_MSG(false);
+		return;
+	}
+
+	__ASSERT_NO_MSG(cb);
+
+	bt_request_cb = cb;
+}
+
+int fp_storage_ak_bond_conn_create(const void *conn_ctx, const bt_addr_le_t *addr,
+				   const struct fp_account_key *account_key)
+{
+	if (!IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)) {
+		__ASSERT_NO_MSG(false);
+		return -ENOTSUP;
+	}
+
+	if (!is_enabled) {
+		return -EACCES;
+	}
+
+	int err;
+	uint8_t id;
+	struct fp_bond *bond;
+	struct fp_bond bond_rollback_copy = (struct fp_bond){0};
+
+	if (account_key) {
+		err = ak_id_get(&id, account_key);
+		if (err) {
+			return err;
+		}
+	} else {
+		/* The actual ID will be determined during Account Key write. */
+		id = BOND_AK_ID_UNKNOWN;
+	}
+
+	bond = fp_bond_free_get();
+	if (!bond) {
+		return -ENOMEM;
+	}
+
+	bond->conn_ctx = conn_ctx;
+	bond->bonded = false;
+	bond->nv.account_key_metadata_id = id;
+	bond->nv.addr = *addr;
+
+	err = fp_bond_settings_save(bond, &bond_rollback_copy);
+	if (err) {
+		return err;
+	}
+
 	return 0;
 }
 
+void fp_storage_ak_bond_conn_confirm(const void *conn_ctx, const bt_addr_le_t *addr)
+{
+	if (!IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)) {
+		__ASSERT_NO_MSG(false);
+		return;
+	}
+
+	if (!is_enabled) {
+		return;
+	}
+
+	struct fp_bond *bond;
+	struct fp_bond *bond_duplicate;
+	int err;
+
+	bond = fp_bond_get(conn_ctx);
+	if (!bond) {
+		return;
+	}
+
+	bond->bonded = true;
+
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		bond_duplicate = bonded_duplicate_get(conn_ctx, addr);
```

Comment:
Maybe we could also make a common `bond_duplicate_remove(fp_bond *)` function. Passing `NULL` could result in filtering out all of the duplicates for each bond entry.

Reply:
We pass conn here. I don't understand your idea.

Reply:
We get duplicates only to remove them. That's why I suggested to simplify API and just provide a functionality that allows to remove duplicates for a given `fp_bond` entry